### PR TITLE
Add a form to add an app from Git

### PR DESCRIPTION
**Pull request by nakabonne**
**What this PR does / why we need it**:
This PR adds React components to select unregistered applications.
To avoid bloating the PR size, let me separate the API part into another PR.

The newly added components are `UnregisteredApplicationList` and `UnregisteredApplicationFilter`. I would be happy to get feedback from you on the code as well as the UI in detail.

Note that this GIF animation is based on temporary hard-coded dummy data, so you can't reproduce it.

Kapture 2021-12-07 at 16 46 23

**Which issue(s) this PR fixes**:

Ref https://github.com/pipe-cd/pipe/issues/2885

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/app/web/src/api/applications.ts [hunk 1]
```diff
@@ -172,3 +174,11 @@ export const updateDescription = async ({
   req.setDescription(description);
   return apiRequest(req, apiClient.updateApplicationDescription);
 };
+
+export const getUnregisteredApplications = ({
+                               }: ListUnregisteredApplicationsRequest.AsObject): Promise<
```
Comment (khanhtc1202) [inline pkg/app/web/src/api/applications.ts:179]: nits, lint does not wok here 🤔

File: pkg/app/web/src/components/application-form/index.tsx [hunk 2]
```diff
@@ -9,16 +9,23 @@ import {
...
 import * as yup from "yup";
-import { APPLICATION_KIND_TEXT } from "~/constants/application-kind";
-import { UI_TEXT_CANCEL, UI_TEXT_SAVE } from "~/constants/ui-text";
+import { APPLICATION_KIND_TEXT, APPLICATION_KIND_BY_NAME } from "~/constants/application-kind";
+import { UI_TEXT_CANCEL, UI_TEXT_SAVE, UI_TEXT_ADD } from "~/constants/ui-text";
 import { useAppSelector } from "~/hooks/redux";
-import { ApplicationKind } from "~/modules/applications";
+import {
+  ApplicationKind,
+} from "~/modules/applications";
```
Comment (khanhtc1202) [inline pkg/app/web/src/components/application-form/index.tsx:25]:
```suggestion
import { ApplicationKind } from "~/modules/applications";
```
nits

File: pkg/app/web/src/api/applications.ts (hunk 1 above)
Comment (nakabonne) [reply pkg/app/web/src/api/applications.ts:179]: oops

File: pkg/app/web/src/components/applications-page/add-application-drawer/index.tsx [hunk 3]
```diff
@@ -60,6 +64,11 @@ export const AddApplicationDrawer: FC<AddApplicationDrawerProps> = memo(
       }
     }, [onClose, formik]);
 
+    const [showConfirmToAddFromGit, setShowConfirmToAddFromGit] = useState(false);
+    const handleAddFromGit = useCallback(() => {
+      setShowConfirmToAddFromGit(true);
+    }, []);
```
Comment (khanhtc1202) [inline pkg/app/web/src/components/applications-page/add-application-drawer/index.tsx:70]: nits, we may not need to use this `useCallback`

**Comment by khanhtc1202** [comment]
Nice work @nakabonne 👍  for the UI, not sure how does it actually look like on the local but it would be aesthetically a bit better if we have a margin around the ADD FROM GIT tab content

just like we have for the current form
Screen Shot 2021-12-07 at 13 54 12

**Comment by nakabonne** [comment]
>it would be aesthetically a bit better if we have a margin around the ADD FROM GIT tab content

Tottally!

File: pkg/app/web/src/components/applications-page/add-application-drawer/index.tsx (hunk 3 above)
Comment (nakabonne) [reply pkg/app/web/src/components/applications-page/add-application-drawer/index.tsx:70]: I see. No input is given then it has no sense 👍

**Comment by nakabonne** [comment]
@khanhtc1202 After a long battle with trivial CSS stuff, it's done. The GIF animation got updated so check it out!

File: pkg/app/web/src/components/application-form/index.tsx [hunk 4]
```diff
@@ -425,3 +437,211 @@ export const ApplicationForm: FC<ApplicationFormProps> = memo(
...
+      <Box width="100%">
+        <Typography className={classes.title} variant="h6">
+          Select the application to register
+        </Typography>
+        <Divider />
+        <Box width="100%" m={0} px={2}>
+          <UnregisteredApplicationFilter
+            onChange={handleFilterChange}
+          />
+          <Accordion disabled>
```
Comment (khanhtc1202) [inline pkg/app/web/src/components/application-form/index.tsx:557]: nits, add a little space between this and the filters group above make it looks better 😉

Comment (nakabonne) [reply pkg/app/web/src/components/application-form/index.tsx:557]: I feel you. I'm now stuck with that 😵‍💫

File: pkg/app/web/src/components/applications-page/add-application-drawer/index.tsx [hunk 5]
```diff
@@ -72,6 +78,9 @@ export const AddApplicationDrawer: FC<AddApplicationDrawerProps> = memo(
             {...formik}
             title={`Add a new application to "${projectName}" project`}
             onClose={handleClose}
+            onAddFromGit={() => {
+              setShowConfirmToAddFromGit(true);
+            }}
```
Comment (khanhtc1202) [inline pkg/app/web/src/components/applications-page/add-application-drawer/index.tsx:83]:
```suggestion
            onAddFromGit={() => setShowConfirmToAddFromGit(true)}
```
nits, just as in L109

File: pkg/app/web/src/components/application-form/index.tsx (hunk 4 above)
Comment (khanhtc1202) [reply pkg/app/web/src/components/application-form/index.tsx:557]: According the API https://mui.com/api/accordion/#props
we can pass `<Accordion className={classes.xxx}` here and define the padding there to make the CSS affect only this element 👀

Comment (nakabonne) [reply pkg/app/web/src/components/application-form/index.tsx]: Thank you. I applied it!

Comment (nakabonne) [reply pkg/app/web/src/components/application-form/index.tsx]: It looks like:
image

Comment (khanhtc1202) [reply pkg/app/web/src/components/application-form/index.tsx]: Nice thank you 🙌

**Comment by khanhtc1202** [comment]
Nice work, can't wait to see it on prod 🙌 
/lgtm

**Comment by nghialv** [comment]
Great! Can't wait to touch it.

/approve

**Pull request merged**